src/main/java/fi/aalto/cs/apluscourses/utils/Tree.java

```diff
@@ -0,0 +1,38 @@
+package fi.aalto.cs.apluscourses.utils;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.function.Predicate;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public interface Tree {
+  @NotNull
+  List<? extends Tree> getChildren();
+
+  /**
+   * Traverses the tree in depth-first pre-order and finds the first node that satisfies the
```

Comment:
what does `depth-first pre-order` mean? :nerd_face:

Reply:
```
    A
   / \
  B   C
 / \
D   E
```
Depth first search with pre-order for this tree is: A, B, D, E, C.
Compare this to DFS with post-order (where parent nodes are visited after their children): D, E, B, C, A.
There's also breadth first search: A, B, C, D, E.